polynote-kernel/src/main/scala/polynote/kernel/util/Publish.scala

```diff
@@ -0,0 +1,35 @@
+package polynote.kernel.util
+
+import fs2.Sink
+import fs2.concurrent.Topic
+
+/**
+  * Captures only the ability to Publish, as in [[Topic]], but without the ability to subscribe. This means it can
+  * be contravariant.
+  */
+trait Publish[F[_], -T] {
```

Comment:
This is some type wizardry I'm not very familiar with. It looks like the main point of `Publish` is for the contravariance. What is it giving us? (maybe it's later in the PR, I'm not sure I'll be able to spot it though 😕)

Reply:
Ok, so basically we have some function `Thing[B] => Z` somewhere and we want to be able to pass `Thing[A]`s to it (assuming `A >: B`), since we know that at the end of the day we won't do anything special with anything less than `A` (i.e., we'll treat all `A` and below the same). 

I guess that makes sense but I guess I don't still fully grok it: Couldn't we just loosen the restriction of whatever is expecting a `Publish[KernelStatus]`? Why not have `Publish[Message]` everywhere? 

What am I missing?

Reply:
Because you might want to pass the `Publish` to a thing that shouldn't be publishing arbitrary messages, only `KernelStatus` messages. For example.

I think `Topic` itself should probably be broken down into `Publish` and `Subscribe` (much like their `Queue` extends `Enqueue`, which is a useful separation of capabilities... though they didn't make it contravariant. Oh well). That would be a PR for fs2 though.

Reply:
Ah, so we want to be able to specify that some class Foo can only publish `KernelStatus` because we want type safety within the class, but when we tell Foo *how* to publish a `KernelStatus` we want to be able to send in our `Publish[Message]` to it. 

So we get type safety in the specific case (Foo only publishes `KernelStatus`) without sacrificing flexibility in the general case (we don't need special Publishers for each type).